timemachine/ff/handlers/nonbonded.py

```diff
@@ -499,6 +506,118 @@ def static_parameterize(mol):
         return compute_or_load_oe_charges(mol, mode=AM1)
 
 
+@jit
+def eval_nn(params_by_layer, features):
+    def activation(x):
+        return x / (1 + jnp.exp(-x))  # silu
+
+    layers_by_index = list(sorted(params_by_layer.keys()))
+
+    x = features
+    for layer in layers_by_index[:-1]:
+        W = params_by_layer[layer]
+        x = activation(jnp.dot(W, x))
+
+    # last layer skips activation
+    W = params_by_layer[layers_by_index[-1]]
+    return jnp.squeeze(jnp.dot(W, x))  # scalar
+
+
+class NNHandler(SerializableMixIn):
+    def __init__(self, layer_sizes, params, props):
+        assert len(layer_sizes) == 1
+        assert len(params) == 1
+        self.smirks = layer_sizes
+        self.params = np.array(params, dtype=np.float64)
+        self.props = props
+
+    @staticmethod
+    def get_bond_idxs_and_charge_deltas(flat_params, init_mol_charges, encoded_layer_sizes, mol):
+        layer_sizes = pickle.loads(base64.b64decode(encoded_layer_sizes[0]))
+        layer_idxs = list(range(len(layer_sizes)))
+        features = pickle.loads(base64.b64decode(mol.GetProp(NN_FEATURES_PROPNAME)))
+        atom_features = features["atom_features"]
+        bond_idx_features = features["bond_idxs"]
+        bond_src_features = features["bond_src_features"]
+        bond_dst_features = features["bond_dst_features"]
+
+        # symmetrize features
+        sym_charge_idxs = defaultdict(list)
+        for i, q in enumerate(init_mol_charges):
+            sym_charge_idxs[float(q)].append(i)
+
+        # all atoms in the sym. group should have the same features
+        for sym_group in sym_charge_idxs.values():
+            sym_atom_features = np.mean([atom_features[j] for j in sym_group], axis=0)
+            for j in sym_group:
+                atom_features[j, :] = sym_atom_features
+
+        # all bonds in the paired sym. group should have the same features
+        bond_features_by_sym_idx = defaultdict(list)
+        for i, bond_idx in enumerate(bond_idx_features):
+            bond_feature = np.concatenate([bond_src_features[i], bond_dst_features[i]])
+            q0 = float(init_mol_charges[bond_idx[0]])
+            q1 = float(init_mol_charges[bond_idx[1]])
+            bond_features_by_sym_idx[(q0, q1)].append(bond_feature)
```

Comment:
Not sure how to think about the bond src/dst features.

(By default, "undirected" bond features seem simplest to me, but the implemented approach seems potentially more expressive: it expects "directed, antisymmetric, paired" bond features. Depending on the model, maybe these directed bond features make the atom features redundant.)

Reply:
Possibly, not sure. I wanted to do a first pass and include both, but we can try different options.